File: python/problem137.py
from typing import Sequence


class Solution:
    def singleNumber(self, nums: Sequence[int]) -> int:
        s = set(nums)
        return (sum(s) * 3 - sum(nums)) // 2

        # Consider a generalization
        # "Given an array of integers, every element appears k (k > 1) times except for one,
        # which appears p times (p >= 1, p % k != 0). Find that single one."

File: cpp/problem137.cpp
#include <unordered_set>
#include <vector>
using namespace std;

class Solution {
   public:
    int singleNumber(vector<int>& nums) {
        // unordered_set<int> s;
        // long sum = 0;
        // for (int i : nums) {
        //     s.insert(i);
        //     sum += i;
        // }
        // for (long i : s) {
        //     sum -= 3 * i;
        // }
        // return -sum / 2;

        int ones = 0, twos = 0;
        for (int i : nums){
            ones = (ones ^ i) & ~twos;
            twos = (twos ^ i) & ~ones;
        }
        return ones;
    }
};